Implement Observable#drop

maraisr opened the pull request:
#### f1556cc6daef8bf209e079fa87369c76767fbf01
<pre>
Implement Observable#drop
<a href="https://bugs.webkit.org/show_bug.cgi?id=277224">https://bugs.webkit.org/show_bug.cgi?id=277224</a>

Reviewed by Darin Adler.

This Pull Request implements the `drop` method for the `Observable`
class. This is required by the specification, for
[`#drop`](<a href="https://wicg.github.io/observable/#dom-observable-drop).">https://wicg.github.io/observable/#dom-observable-drop).</a> Also adds
the `InternalObserverDrop` and `SubscriberCallbackDrop` to support this.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-drop.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-drop.any.worker-expected.txt: Rebaselined.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverDrop.cpp: Added.
(WebCore::createSubscriberCallbackDrop):
* Source/WebCore/dom/InternalObserverDrop.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::drop):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:

Canonical link: <a href="https://commits.webkit.org/281734@main">https://commits.webkit.org/281734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50dd5c0c24495d5a26f0241cc53e69fdd692f693

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64648 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11264 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62847 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11539 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49094 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7811 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37321 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29925 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34009 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9829 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10177 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55851 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10127 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66377 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4661 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9959 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56462 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4682 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52571 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56641 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13565 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3860 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35881 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36963 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38056 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36708 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/InternalObserverDrop.cpp:
```diff
@@ -0,0 +1,132 @@
+        }
+
+    private:
+        Ref<Observable> m_sourceObservable;
```
Traditionally in WebKit code we put the data members last after all the function members. I hadn’t noticed this in your other pull requests.

keithamus replied:
I'll follow up with a PR to clean all of these up.

maraisr replied:
I was actually going to do it too 🙈 opportunity to pair?

maraisr committed:
Implement Observable#drop
https://bugs.webkit.org/show_bug.cgi?id=277224

Reviewed by Darin Adler.

This Pull Request implements the `drop` method for the `Observable`
class. This is required by the specification, for
[`#drop`](https://wicg.github.io/observable/#dom-observable-drop). Also adds
the `InternalObserverDrop` and `SubscriberCallbackDrop` to support this.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-drop.any-expected.txt: Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-drop.any.worker-expected.txt: Rebaselined.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverDrop.cpp: Added.
(WebCore::createSubscriberCallbackDrop):
* Source/WebCore/dom/InternalObserverDrop.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::drop):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:

Canonical link: https://commits.webkit.org/281734@main
